core/finality-grandpa/src/lib.rs

```diff
@@ -0,0 +1,767 @@
+	Blockchain(String),
+	/// A timer failed to fire.
+	Timer(::tokio::timer::Error),
+}
+
+impl From<GrandpaError> for Error {
+	fn from(e: GrandpaError) -> Self {
+		Error::Grandpa(e)
+	}
+}
+
+/// A handle to the network. This is generally implemented by providing some
+/// handle to a gossip service or similar.
+///
+/// Intended to be a lightweight handle such as an `Arc`.
+pub trait Network: Clone {
+	/// A stream of input messages for a topic.
+	type In: Stream<Item=Vec<u8>,Error=()>;
+
+	/// Get a stream of messages for a specific round. This stream should
+	/// never logically conclude.
+	fn messages_for(&self, round: u64) -> Self::In;
+
+	/// Send a message at a specific round out.
+	fn send_message(&self, round: u64, message: Vec<u8>);
+
+	/// Clean up messages for a round.
+	fn drop_messages(&self, round: u64);
+}
+
+/// Something which can determine if a block is known.
+pub trait BlockStatus<Block: BlockT> {
+	/// Return `Ok(Some(number))` or `Ok(None)` depending on whether the block
+	/// is definitely known and has been imported.
+	/// If an unexpected error occurs, return that.
+	fn block_number(&self, hash: Block::Hash) -> Result<Option<u32>, Error>;
+}
+
+impl<B, E, Block: BlockT> BlockStatus<Block> for Arc<Client<B, E, Block>> where
+	B: Backend<Block, Blake2Hasher>,
+	E: CallExecutor<Block, Blake2Hasher>,
+	NumberFor<Block>: As<u32>,
+{
+	fn block_number(&self, hash: Block::Hash) -> Result<Option<u32>, Error> {
+		self.block_number_from_id(&BlockId::Hash(hash))
+			.map_err(|e| Error::Blockchain(format!("{:?}", e)))
+			.map(|num| num.map(|n| n.as_()))
+	}
+}
+
+/// Buffering imported messages until blocks with given hashes are imported.
+pub struct UntilImported<Block: BlockT, Status, I> {
+	import_notifications: Fuse<ImportNotifications<Block>>,
+	status_check: Status,
+	inner: Fuse<I>,
+	ready: VecDeque<SignedMessage<Block>>,
+	check_pending: Interval,
+	pending: HashMap<Block::Hash, Vec<SignedMessage<Block>>>,
+}
+
+impl<Block: BlockT, Status: BlockStatus<Block>, I: Stream> UntilImported<Block, Status, I> {
+	fn new(
+		import_notifications: ImportNotifications<Block>,
+		status_check: Status,
+		stream: I,
+	) -> Self {
+		// how often to check if pending messages that are waiting for blocks to be
+		// imported can be checked.
+		//
+		// the import notifications interval takes care of most of this; this is
+		// used in the event of missed import notifications
+		const CHECK_PENDING_INTERVAL: Duration = Duration::from_secs(5);
+		let now = Instant::now();
+
+		let check_pending = Interval::new(now + CHECK_PENDING_INTERVAL, CHECK_PENDING_INTERVAL);
+		UntilImported {
+			import_notifications: import_notifications.fuse(),
+			status_check,
+			inner: stream.fuse(),
+			ready: VecDeque::new(),
+			check_pending,
+			pending: HashMap::new(),
+		}
+	}
+}
+
+impl<Block: BlockT, Status: BlockStatus<Block>, I> Stream for UntilImported<Block, Status, I>
+	where I: Stream<Item=SignedMessage<Block>,Error=Error>
```

Comment:
(nitpick) would space after comma be appropriate?

Reply:
i don't think so, it's not typical for these kinds of bounds